crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -131,46 +131,78 @@ impl CommandQueue {
+                let meta = unsafe { cursor.cast::<CommandMeta>().read_unaligned() };
+                // Advance to the bytes just after `meta`, which represent a type-erased command.
+                // SAFETY: For most types of `Command`, the pointer immediately following the metadata
+                // is guaranteed to be in bounds. If the command is a zero-sized type (ZST), then the cursor
+                // might be 1 byte past the end of the buffer, which is safe.
+                cursor = unsafe { cursor.add(std::mem::size_of::<CommandMeta>()) };
+                // Construct an owned pointer to the command.
+                // SAFETY: It is safe to transfer ownership out of `self.bytes`, since the call to `set_len(0)` above
+                // guarantees that nothing stored in the buffer will get observed after this function ends.
+                // `cmd` points to a valid address of a stored command, so it must be non-null.
+                let cmd = unsafe {
+                    OwningPtr::<Unaligned>::new(std::ptr::NonNull::new_unchecked(cursor.cast()))
+                };
+                // SAFETY: The data underneath the cursor must correspond to the type erased in metadata,
+                // since they were stored next to each other by `.push()`.
+                // For ZSTs, the type doesn't matter as long as the pointer is non-null.
+                let size = unsafe { (meta.consume_command_and_get_size)(cmd, &mut world) };
+                // Advance the cursor past the command. For ZSTs, the cursor will not move.
+                // At this point, it will either point to the next `CommandMeta`,
+                // or the cursor will be out of bounds and the loop will end.
+                // SAFETY: The address just past the command is either within the buffer,
+                // or 1 byte past the end, so this addition will not overflow the pointer's allocation.
+                cursor = unsafe { cursor.add(size) };
+
+                if let Some(world) = &mut world {
+                    // Mark world as flushing so calls to `World::flush_commands` don't cause multiple queues to be applying simultaneously
+                    world.flushing_commands = true;
+                    // If the command we just applied generated more commands we must apply those first
+                    if !world.command_queue.is_empty() {
+                        // If our current list of commands isn't complete push it to the `resolving_commands` stack to be applied after
+                        if cursor < end {
+                            resolving_commands.push((cursor, end, bytes));
+                        }
+                        // Set our variables to start applying the new queue
+                        bytes = std::mem::take(&mut world.command_queue.bytes);
+                        let bytes_range = bytes.as_mut_ptr_range();
+                        cursor = bytes_range.start;
+                        end = bytes_range.end;
```

Comment:
If the world's command queue already had commands on it before this function got called, won't that result in the command getting applied out of order? Should we add the world's command queue to the front of `resolving_commands`?

Reply:
Yes, we should be applying from the internal queue first. I had the same thought when replying to your questions above.